phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -3199,12 +3188,9 @@ public Void call() throws Exception {
                                     }
                                     return null;
                                 } catch (UpgradeRequiredException e) {
-                                    // This will occur in 3 cases:
-                                    // 1. SYSTEM.CATALOG does not exist and we don't want to allow the user to create it i.e.
-                                    //    !isAutoUpgradeEnabled or isDoNotUpgradePropSet is set
-                                    // 2. SYSTEM.CATALOG exists and its timestamp < MIN_SYSTEM_TABLE_TIMESTAMP
-                                    // 3. SYSTEM.CATALOG exists, but client and server-side namespace mapping is enabled so
-                                    //    we need to migrate SYSTEM tables to the SYSTEM namespace
+                                    // This will occur when SYSTEM.CATALOG exists, but client and
```

Comment:
So if the client has `doNotUpgrade == true`, SYSTEM.CATALOG exists, server and client both have ns-mapping enabled -> this will continue throwing `UpgradeRequiredException`?
What about the other 2 cases where we used to fall into this block and later `setUpgradeRequired`?

For original case1 (syscat doesn't exist and client has doNotUpgrade == true) -> You changes will lead to a TableNotFoundException and as mentioned earlier, can we catch and rethrow as UpgradeRequiredException so that this follows the original code path?

What about case2?

Reply:
We are not checking case 2 if DoNotUpgrade conf is set to true.